C/services/core/configuration_manager.cpp

```diff
@@ -0,0 +1,512 @@
+	ResultSet::Row* row = *it;
+	if (!row)
+	{
+		delete categoryItems;
+		throw NoSuchRowxception();
+	}	
+
+	ResultSet::ColumnValue* key = row->getColumn("key");
+	ResultSet::ColumnValue* description = row->getColumn("description");
+	ResultSet::ColumnValue* items = row->getColumn("value");
+
+	// Create string representation of JSON object
+	rapidjson::StringBuffer buffer;
+	rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+	const rapidjson::Value *v = items->getJSON();
+	v->Accept(writer);
+
+	const string sItems(buffer.GetString(), buffer.GetSize());
+
+	// Create category object
+	ConfigCategory theVal(key->getString(), sItems);
+
+	// Set description
+	theVal.setDescription(description->getString());
+
+	delete categoryItems;
+
+	return theVal;
+}
+
+/**
+ * Create or update a new category
+ *
+ * @param categoryName		The category name
+ * @param categoryDescription	The category description
+ * @param categoryItems		The category items
+ * @return			The ConfigCategory object
+ *				with "value" and "default"
+ *				of the new category added
+ *				or the merged configuration
+ *				of the updated confguration.
+ */
+
+ConfigCategory ConfigurationManager::createCategory(const std::string& categoryName,
+						    const std::string& categoryDescription,
+						    const std::string& categoryItems) const
+{
+	// Fill the ready to insert category object with input data
+	ConfigCategory preparedValue(categoryName, categoryItems);
+	preparedValue.setDescription(categoryDescription);
+
+	try
+	{
+		// Abort if items contain both value and default
+		preparedValue.checkDefaultValuesOnly();
+
+		// Add 'value' from 'default' for each item
+		preparedValue.setItemsValueFromDefault();
+	}
+	catch (ConfigMalformed* e)
+	{
+		delete e;
+		throw;
+	}
+	catch (ConfigValueFoundWithDefault* e)
+	{
+		delete e;
+		throw;
+	}
+	catch (...)
+	{
+		throw;
+	}
+
+	// Parse JSON input
+	Document doc;
+	// Parse the prepared input category with "value" and "default"
+	doc.Parse(preparedValue.itemsToJSON().c_str());
+	if (doc.HasParseError())
+	{
+		throw new ConfigMalformed();
+	}
+
+	// Set the JSON string for merged category values
+	string updatedItems;
+
+	// SELECT * FROM foglamp.configuration WHERE key = categoryName
+	const Condition conditionKey(Equals);
+	Where *wKey = new Where("key", conditionKey, categoryName);
+	Query qKey(wKey);
+
+	// Query via storage client
+	ResultSet* result = m_storage->queryTable("configuration", qKey);
+	if (!result->rowCount())
+	{
+		// Prepare insert values for insertTable
+		InsertValues newCategory;
+		newCategory.push_back(InsertValue("key", categoryName));
+		newCategory.push_back(InsertValue("description", categoryDescription));
+		// Set "value" field for inseert using the JSON document object
+		newCategory.push_back(InsertValue("value", doc));
+
+		// Do the insert
+		if (!m_storage->insertTable("configuration", newCategory))
+		{
+			delete result;
+			throw DBOperationException();
```

Comment:
Can we make it a StorageOperationException? We might not have a database as such in future.

Reply:
Done